### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by danikkrapivnitskiy**

### CODE REVIEW

File: TSTutorial/homeworks/hw_4/Types.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { OrderStatusEnum, PaymentMethodEnum } from "./Enums"
+
+export type TOrder = {
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Types.ts]: Better use interfaces for objects

File: TSTutorial/homeworks/hw_4/Pizzeria.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        console.log(`Welcome to Pizzeria '${this.name}' on adress '${this.adress}' at '${this.workingHours}'`)
+        this.orderService = new OrderService()
+    }
+
+    addOrders(order: TOrder[]): void {
+        this.orderService.addOrder(order)
+        if (this.orderService.getOrderByStatus(OrderStatusEnum.PROCESSING) !== null
+            && this.orderService.getOrderByStatus(OrderStatusEnum.IN_PROGRESS) === null) {
+            this.orderService.setStatusToOrder(OrderStatusEnum.PROCESSING, OrderStatusEnum.IN_PROGRESS)
+        } else this.orderService.setStatusToOrder(OrderStatusEnum.PROCESSING, OrderStatusEnum.PENDING)
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Pizzeria.ts]: This implementation should be a private part of OrderService class. Pizzeria class should not have any knowledge about how it works under cover, and must rely only on the provided inverface

File: TSTutorial/homeworks/hw_4/Pizzeria.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    prepareOrder(): void {
+        this.orderService.setStatusToOrder(OrderStatusEnum.IN_PROGRESS, OrderStatusEnum.COMPLETED)
+        if (this.orderService.getOrderByStatus(OrderStatusEnum.PROCESSING) !== null) {
+            this.orderService.setStatusToOrder(OrderStatusEnum.PROCESSING, OrderStatusEnum.IN_PROGRESS)
+        } else if (this.orderService.getOrderByStatus(OrderStatusEnum.PENDING) !== null) {
+            this.orderService.setStatusToOrder(OrderStatusEnum.PENDING, OrderStatusEnum.IN_PROGRESS)
+        }
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Pizzeria.ts]: This implementation should be a private part of OrderService class. Pizzeria class should not have any knowledge about how it works under cover, and must rely only on the provided inverface

File: TSTutorial/homeworks/hw_4/Meals.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    calculatePrice(): number {
+        switch (this.cup) {
+            case CupSizeEnum.SMALL:
+                return 10 * this.quantity;
+            case CupSizeEnum.LARGE:
+                return 20 * this.quantity;
+            default:
+                return 15 * this.quantity;
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Meals.ts:26]: Better to store such data in "database", or in json/js file in our case

File: TSTutorial/homeworks/hw_4/Meals.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        switch (this.doughType) {
+          case DoughTexturePizzaEnum.Crispy: 
+            basePrice += 10;
+            break;
+          case DoughTexturePizzaEnum.Chewy:
+            basePrice += 5;
+            break;
+          case DoughTexturePizzaEnum.Tough:
+            basePrice -= 5;
+            break;
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Meals.ts:48]: Better to store such data in "database", or in json/js file in our case

File: TSTutorial/homeworks/hw_4/Meals.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+}
+
+export class Pizza extends Meal {
+    constructor(private doughType: DoughTexturePizzaEnum, private toppings: string[] = [], private size: PizzaSizeEnum) {
+        super()
+    }
+
+    calculatePrice(): number {
+        let basePrice = 100;
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Meals.ts:37]: Better to store such data in "database", or in json/js file in our case

File: TSTutorial/homeworks/hw_4/Order.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import {TOrder, IMeal} from "./Types"
+import {OrderStatusEnum, PaymentMethodEnum} from "./Enums";
+
+export class Order {
```
Comment (anatoly-karpovich) [inline TSTutorial/homeworks/hw_4/Order.ts]: update this class to be used in OrderService, and implement interface for services like OrderService to depend on

**Review by anatoly-karpovich — APPROVED** [review]
Well done!

File: TSTutorial/homeworks/hw_4/Types.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+    id: number;
+    customerId: number;
+    customerName: string;
+    order: Object[];
```
Comment (VladimirLQA) [inline TSTutorial/homeworks/hw_4/Types.ts]: Not good way to use `Object` as type
The same for `IMeal.getMealInfo()`

File: TSTutorial/homeworks/hw_4/Order.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        ? this.orders.find(entity => entity.customerName === filter) 
+        : this.orders.find(entity => entity.customerId === filter) 
+        if (order === undefined) {
+            throw new Error(`Order for customer ${filter} not found`)
+        }
+        return order;
+    }
+
+    private verifyIfCustomerExists(customerName: string): boolean {
+        return this.orders.filter(orderItem => orderItem.customerName === customerName).length > 0;
```
Comment (VladimirLQA) [inline TSTutorial/homeworks/hw_4/Order.ts:59]: Consider to use `some` method

File: TSTutorial/homeworks/hw_4/Order.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import {TOrder, IMeal} from "./Types"
+import {OrderStatusEnum, PaymentMethodEnum} from "./Enums";
+
+export class Order {
+    private customerId: number = 0 
+    private orders: TOrder[] = [] 
+    private id: number = 1
+    private counter: number = 0
+
+    makeOrder(customerName: string, meal: IMeal[] | IMeal, paymentMethod: PaymentMethodEnum) {
```
Comment (VladimirLQA) [inline TSTutorial/homeworks/hw_4/Order.ts:9]: You can separate this method like
```ts
makeOrder () {
if (Array.isArray(meal)) {
            meal.forEach(item => this.processMeal(customerName, item, paymentMethod));
   } else { 
      this.processMeal(customerName, item, paymentMethod))
  } 
}

private processMeal() {
 // with logic that remained
}

```

**Review by VladimirLQA — APPROVED** [review]

**Pull request merged**